File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/queue/Queue.kt
package com.margelo.nitro.nitroplayer.queue

import com.margelo.nitro.nitroplayer.TrackItem
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Queue class that manages a list of tracks.
 * Thread-safe implementation using CopyOnWriteArrayList.
 */
class Queue {
    private val tracks: MutableList<TrackItem> = CopyOnWriteArrayList()

    /**
     * Get all tracks in the queue
     */
    fun getTracks(): List<TrackItem> = ArrayList(tracks)

    /**
     * Get tracks as an array (for compatibility with existing API)
     */
    fun getTracksArray(): Array<TrackItem> = tracks.toTypedArray()

    /**
     * Add a single track to the queue
     */
    fun addTrack(track: TrackItem) {
        tracks.add(track)
    }

    /**
     * Add a track at a specific index
     */
    fun addTrackAtIndex(
        track: TrackItem,
        index: Int,
    ) {
        if (index < 0 || index > tracks.size) {
            tracks.add(track)
        } else {
            tracks.add(index, track)
        }
    }

    /**
     * Load multiple tracks into the queue (replaces existing queue)
     */
    fun loadTracks(newTracks: List<TrackItem>) {
        tracks.clear()
        tracks.addAll(newTracks)
    }

    /**
     * Remove a track by ID
     * @return true if track was found and removed, false otherwise
     */
    fun removeTrack(id: String): Boolean = tracks.removeAll { it.id == id }

    /**
     * Clear all tracks from the queue
     */
    fun clear() {
        tracks.clear()
    }

    /**
     * Get the size of the queue
     */
    fun size(): Int = tracks.size

    /**
     * Check if the queue is empty
     */
    fun isEmpty(): Boolean = tracks.isEmpty()

    /**
     * Get a track by index
     */
    fun getTrack(index: Int): TrackItem? =
        if (index >= 0 && index < tracks.size) {
            tracks[index]
        } else {
            null
        }

    /**
     * Get a track by ID
     */
    fun getTrackById(id: String): TrackItem? = tracks.find { it.id == id }

    /**
     * Get the index of a track by ID
     */
    fun getTrackIndex(id: String): Int = tracks.indexOfFirst { it.id == id }
}

File: react-native-nitro-player/ios/queue/Queue.swift
//
//  Queue.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 24/11/25.
//

import Foundation

/// Queue class that manages a list of tracks.
/// Thread-safe implementation using a serial dispatch queue.
class Queue {
  private var tracks: [TrackItem] = []
  private let queue = DispatchQueue(label: "com.margelo.nitro.nitroplayer.queue")

  /**
   * Get all tracks in the queue
   */
  func getTracks() -> [TrackItem] {
    return queue.sync {
      return Array(tracks)
    }
  }

  /**
   * Add a single track to the queue
   */
  func addTrack(_ track: TrackItem) {
    queue.sync {
      self.tracks.append(track)
    }
  }

  /**
   * Add a track at a specific index
   */
  func addTrackAtIndex(_ track: TrackItem, index: Int) {
    queue.sync {
      if index < 0 || index > self.tracks.count {
        self.tracks.append(track)
      } else {
        self.tracks.insert(track, at: index)
      }
    }
  }

  /**
   * Load multiple tracks into the queue (replaces existing queue)
   */
  func loadTracks(_ newTracks: [TrackItem]) {
    queue.sync {
      self.tracks.removeAll()
      self.tracks.append(contentsOf: newTracks)
    }
  }

  /**
   * Remove a track by ID
   * @return true if track was found and removed, false otherwise
   */
  func removeTrack(id: String) -> Bool {
    return queue.sync {
      let initialCount = self.tracks.count
      self.tracks.removeAll { $0.id == id }
      return self.tracks.count < initialCount
    }
  }

  /**
   * Clear all tracks from the queue
   */
  func clear() {
    queue.sync {
      self.tracks.removeAll()
    }
  }

  /**
   * Get the size of the queue
   */
  func size() -> Int {
    return queue.sync {
      return self.tracks.count
    }
  }

  /**
   * Check if the queue is empty
   */
  func isEmpty() -> Bool {
    return queue.sync {
      return self.tracks.isEmpty
    }
  }

  /**
   * Get a track by index
   */
  func getTrack(index: Int) -> TrackItem? {
    return queue.sync {
      if index >= 0 && index < self.tracks.count {
        return self.tracks[index]
      } else {
        return nil
      }
    }
  }

  /**
   * Get a track by ID
   */
  func getTrackById(id: String) -> TrackItem? {
    return queue.sync {
      return self.tracks.first { $0.id == id }
    }
  }

  /**
   * Get the index of a track by ID
   */
  func getTrackIndex(id: String) -> Int {
    return queue.sync {
      return self.tracks.firstIndex { $0.id == id } ?? -1
    }
  }
}
